src/fdb5/database/Key.cc

```diff
@@ -473,6 +365,207 @@ std::string Key::toString() const {
     return res;
 }
 
+
 //----------------------------------------------------------------------------------------------------------------------
 
+Key::Key() :
+    BaseKey() {}
+
+// Key::Key(const std::string &s) :
+//     BaseKey({}) {
+
+//     eckit::Tokenizer parse(":", true);
+//     eckit::StringList values;
+//     parse(s, values);
+// }
+
+Key::Key(const eckit::StringDict &keys) :
+    BaseKey(keys) {}
+
+Key::Key(eckit::Stream& s) :
+    BaseKey() {
+    decode(s);
+}
+
+Key::Key(std::initializer_list<std::pair<const std::string, std::string>> l) :
+    BaseKey(l) {}
+
+Key Key::parseString(const std::string& s) {
+
+    eckit::Tokenizer parse1(",");
+    eckit::Tokenizer parse2("=");
+    eckit::StringDict keys;
+
+    eckit::StringList v;
+    parse1(s, v);
+    for (const auto& bit : v) {
+        eckit::StringList kv;
+        parse2(bit, kv);
+        ASSERT(kv.size() == 2);
+        keys.emplace(std::move(kv[0]), std::move(kv[1]));
+    }
+
+    return Key{keys};
+}
+
+std::string Key::canonicalise(const std::string& keyword, const std::string& value) const {
+    return value;
+}
+
+std::string Key::type(const std::string& keyword) const {
+    return "";
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+// TypedKey::TypedKey(const Key& key) : 
+//     BaseKey(key), registry_(nullptr) {}
+
+TypedKey::TypedKey(const Key& key, const std::shared_ptr<TypesRegistry> reg) : 
+    BaseKey(key), registry_(reg) {}
+
+TypedKey::TypedKey(const std::shared_ptr<TypesRegistry> reg) :
+    BaseKey({}), registry_(reg) {}
+
+TypedKey::TypedKey(const std::string &s, const Rule *rule) :
+    BaseKey({}), registry_(rule ? rule->registry() : nullptr) {
```

Comment:
need more attention here
contradicting to have a conditional `rule` here while later in the body there is an `ASSERT(rule)`
`if (rule) { rule->fill }` maybe or `rule` is a ref ?